File: src/main/java/mods/orca/mffs/blocks/base/tile/TileMachine.java
package mods.orca.mffs.blocks.base.tile;

import ic2.api.classic.item.IMachineUpgradeItem;
import ic2.api.classic.tile.IMachine;
import ic2.api.energy.event.EnergyTileLoadEvent;
import ic2.api.energy.event.EnergyTileUnloadEvent;
import ic2.api.energy.tile.IEnergyEmitter;
import ic2.api.energy.tile.IEnergySink;
import mods.orca.mffs.MFFSMod;
import net.minecraft.block.state.IBlockState;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.network.EnumPacketDirection;
import net.minecraft.network.NetworkManager;
import net.minecraft.network.play.server.SPacketUpdateTileEntity;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import net.minecraftforge.common.MinecraftForge;

import javax.annotation.Nullable;
import java.util.Set;

public abstract class TileMachine extends TileEntity implements IMachine, IEnergySink {
    private double energyStored = 0;
    public final double maxEnergy;

    private boolean hasChanged = false;
    protected boolean active = false;

    private static final String ENERGY_STORED_KEY = "energyStored";

    public TileMachine(double maxEnergy) {
        super();
        this.maxEnergy = maxEnergy;
    }

    @Override
    public NBTTagCompound writeToNBT(NBTTagCompound compound) {
        compound.setDouble(ENERGY_STORED_KEY, energyStored);
        return super.writeToNBT(compound);
    }

    @Override
    public void readFromNBT(NBTTagCompound compound) {
        energyStored = compound.getDouble(ENERGY_STORED_KEY);
        super.readFromNBT(compound);
    }

    @Nullable
    @Override
    public SPacketUpdateTileEntity getUpdatePacket() {
        if (!hasChanged) {
            return null;
        }

        hasChanged = false;
        return new SPacketUpdateTileEntity(getPos(), getBlockMetadata(), getUpdateTag());
    }

    @Override
    public NBTTagCompound getUpdateTag() {
        NBTTagCompound compound = super.getUpdateTag();
        compound.setDouble(ENERGY_STORED_KEY, energyStored);

        return compound;
    }

    @Override
    public void onDataPacket(NetworkManager net, SPacketUpdateTileEntity pkt) {
        if (net.getDirection() == EnumPacketDirection.CLIENTBOUND) {
            // the server should never be receiving power updates from a client
            // only the client will get updates from the server.
            NBTTagCompound compound = pkt.getNbtCompound();
            if (compound.hasKey(ENERGY_STORED_KEY)) {
                energyStored = compound.getDouble(ENERGY_STORED_KEY);
            }
        }
    }

    @Override
    public double getEnergy() {
        return energyStored;
    }

    @Override
    public boolean useEnergy(double amount, boolean simulate) {
        boolean canDoThat = energyStored - amount >= 0;

        if (!simulate && canDoThat) {
            changeEnergy(energyStored - amount);
        }

        return canDoThat;
    }

    @Override
    public void setRedstoneSensitive(boolean active) {}

    @Override
    public boolean isRedstoneSensitive() {
        return true;
    }

    @Override
    public boolean isProcessing() {
        return false;
    }

    @Override
    public boolean isValidInput(ItemStack par1) {
        return false;
    }

    @Override
    public Set<IMachineUpgradeItem.UpgradeType> getSupportedTypes() {
        return null;
    }

    @Override
    public World getMachineWorld() {
        return world;
    }

    @Override
    public BlockPos getMachinePos() {
        return pos;
    }

    @Override
    public double injectEnergy(EnumFacing directionFrom, double amount, double voltage) {
        changeEnergy(energyStored + amount);
        return Math.max(energyStored - maxEnergy, 0);
    }

    @Override
    public double getDemandedEnergy() {
        return Math.max(maxEnergy - energyStored, 0);
    }

    protected void changeEnergy(double newValue) {
        energyStored = newValue;
        hasChanged = true;
    }

    @Override
    public boolean acceptsEnergyFrom(IEnergyEmitter emitter, EnumFacing side) {
        return true;
    }

    @Override
    public void invalidate() {
        logLoadedTest(false);
        MinecraftForge.EVENT_BUS.post(new EnergyTileUnloadEvent(this));
        super.invalidate();
    }

    @Override
    public void onChunkUnload() {
        logLoadedTest(false);
        MinecraftForge.EVENT_BUS.post(new EnergyTileUnloadEvent(this));
        super.onChunkUnload();
    }

    @Override
    public void onLoad() {
        logLoadedTest(true);
        MinecraftForge.EVENT_BUS.post(new EnergyTileLoadEvent(this));
    }

    @Override
    public boolean shouldRefresh(World world, BlockPos pos, IBlockState oldState, IBlockState newSate) {
        return oldState.getBlock() != newSate.getBlock();
    }

    private void logLoadedTest(boolean l) {
        MFFSMod.logger.info((l ? "loaded" : "unloaded") + (world.isRemote ? " on client" : " on server"));
    }
}

File: src/main/java/mods/orca/mffs/blocks/base/tile/TileMachine.kt
package mods.orca.mffs.blocks.base.tile

import ic2.api.classic.item.IMachineUpgradeItem
import ic2.api.classic.tile.IMachine
import ic2.api.energy.event.EnergyTileLoadEvent
import ic2.api.energy.event.EnergyTileUnloadEvent
import ic2.api.energy.tile.IEnergySink
import mods.orca.mffs.utils.getBooleanOrNull
import mods.orca.mffs.utils.getDoubleOrNull
import net.minecraft.block.state.IBlockState
import net.minecraft.nbt.NBTTagCompound
import net.minecraft.network.EnumPacketDirection
import net.minecraft.network.NetworkManager
import net.minecraft.network.play.server.SPacketUpdateTileEntity
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.EnumFacing
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import net.minecraftforge.common.MinecraftForge
import kotlin.math.max

/**
 * A machine which stores a quantity of energy, which can be used to perform operations.
 *
 * @param maxEnergy The maximum amount of energy this machine can store.
 */
abstract class TileMachine(val maxEnergy: Double) : TileEntity(), IMachine, IEnergySink {

    /**
     * The amount of energy currently stored in this machine.
     */
    private var energy = 0.0

    /**
     * Whether this machine is sensitive to redstone current.
     */
    private var isRedstoneSensitive = false

    /**
     * Whether this machine is currently processing, aka, is currently active/enabled.
     */
    protected var processing = false
        private set

    /**
     * Set whether the machine is processing something, i.e., is active. This method additionally marks the tile as
     * dirty so that it can be synchronised as needed.
     *
     * @param processing Whether the machine is now processing.
     */
    protected fun setProcessing(processing: Boolean) {
        this.processing = processing
        markDirty()
    }

    override fun writeToNBT(compound: NBTTagCompound): NBTTagCompound {

        compound.setDouble(NBT_KEY_ENERGY, energy)
        compound.setBoolean(NBT_KEY_REDSTONE_SENSITIVE, isRedstoneSensitive)
        compound.setBoolean(NBT_KEY_IS_PROCESSING, processing)

        return super.writeToNBT(compound)

    }

    override fun readFromNBT(compound: NBTTagCompound) {

        super.readFromNBT(compound)

        compound.getDoubleOrNull(NBT_KEY_ENERGY)?.let { energy = it }
        compound.getBooleanOrNull(NBT_KEY_REDSTONE_SENSITIVE)?.let { isRedstoneSensitive = it }
        compound.getBooleanOrNull(NBT_KEY_IS_PROCESSING)?.let { processing = it }

    }

    override fun getUpdatePacket(): SPacketUpdateTileEntity? {
        return SPacketUpdateTileEntity(pos, blockMetadata, updateTag)
    }

    override fun getUpdateTag(): NBTTagCompound {
        return super.getUpdateTag().apply(::writeToNBT)
    }

    override fun onDataPacket(net: NetworkManager, pkt: SPacketUpdateTileEntity) {

        when (net.direction) {

            EnumPacketDirection.CLIENTBOUND -> {

                // the server should never be receiving power updates from a client
                // only the client will get updates from the server.
                readFromNBT(pkt.nbtCompound)

            }

            EnumPacketDirection.SERVERBOUND -> {

            }

            else -> Unit

        }

    }

    override fun useEnergy(amount: Double, simulate: Boolean): Boolean {

        val newEnergy = energy - amount

        if (newEnergy < 0) {
            return false
        }

        if (!simulate) {
            changeEnergy(newEnergy)
        }

        return true

    }

    override fun getEnergy() =
        energy

    override fun setRedstoneSensitive(active: Boolean) {
        isRedstoneSensitive = active
        markDirty()
    }

    override fun isRedstoneSensitive() =
        isRedstoneSensitive

    override fun isProcessing() =
        processing

    override fun getMachineWorld(): World =
        world

    override fun getMachinePos(): BlockPos =
        pos

    /**
     * We don't support any item-based upgrades as we use adjacent blocks for modifying behaviour!
     */
    override fun getSupportedTypes() = emptySet<IMachineUpgradeItem.UpgradeType>()

    override fun injectEnergy(directionFrom: EnumFacing, amount: Double, voltage: Double): Double {
        changeEnergy(energy + amount)
        return max(energy - maxEnergy, 0.0)
    }

    override fun getDemandedEnergy(): Double {
        return max(maxEnergy - energy, 0.0)
    }

    /**
     * Modify the amount of energy stored in this tile, marking it for re-saving.
     *
     * @param newValue The new amount of energy to be stored.
     */
    private fun changeEnergy(newValue: Double) {
        energy = newValue
        markDirty()
    }

    override fun invalidate() {
        super.invalidate()
        MinecraftForge.EVENT_BUS.post(EnergyTileUnloadEvent(this))
    }

    override fun onChunkUnload() {
        super.onChunkUnload()
        MinecraftForge.EVENT_BUS.post(EnergyTileUnloadEvent(this))
    }

    override fun onLoad() {
        MinecraftForge.EVENT_BUS.post(EnergyTileLoadEvent(this))
    }

    override fun shouldRefresh(world: World, pos: BlockPos, oldState: IBlockState, newState: IBlockState): Boolean {
        return oldState.block !== newState.block
    }

    private companion object {
        const val NBT_KEY_ENERGY = "energy"
        const val NBT_KEY_REDSTONE_SENSITIVE = "isRedstoneSensitive"
        const val NBT_KEY_IS_PROCESSING = "isProcessing"
    }

}
